modules/identity-shiro/src/main/java/org/opensearch/identity/shiro/realm/InternalRealm.java

```diff
@@ -0,0 +1,107 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.identity.shiro.realm;
+
+import org.apache.shiro.authc.AuthenticationException;
+import org.apache.shiro.authc.AuthenticationInfo;
+import org.apache.shiro.authc.AuthenticationToken;
+import org.apache.shiro.authc.IncorrectCredentialsException;
+import org.apache.shiro.authc.SimpleAuthenticationInfo;
+import org.apache.shiro.authc.UnknownAccountException;
+import org.apache.shiro.authc.UsernamePasswordToken;
+import org.apache.shiro.authc.pam.UnsupportedTokenException;
+import org.apache.shiro.realm.AuthenticatingRealm;
+import org.apache.shiro.authc.UsernamePasswordToken;
+
+import org.opensearch.identity.StringPrincipal;
+
+import java.util.Objects;
+import java.util.Map;
+
+/**
+ * Internal Realm is a custom realm using the internal OpenSearch IdP
+ *
+ * @opensearch.experimental
+ */
+public class InternalRealm extends AuthenticatingRealm {
```

Comment:
this will be changed to AuthorizingRealm in the next iteration correct?

Reply:
It could be, I'd defer a decision to @scrawfor99 as he has been taking point on authorization  grants & checks.

Reply:
It probably will be but I am not super familiar with the realm handling at this point. I will need to look into what that entails. For now, I would not let it change anything about Peter's PR however since it will serve the same functionality and in this use it is basically just an `IdentityRealm`. Peter and @cwperks know more about Shiro than I do however.

Reply:
Assuming identity sticks with shiro, then this would be switched to `AuthorizingRealm` when permissions and privilege evaluation is ready to go to main. When a realm is an `AuthorizingRealm` then it implements a method called `doGetAuthorizationInfo` which is a method to resolve a subject to a list of permissions.